chainer/functions/activation/lstm.py

```diff
@@ -65,49 +65,64 @@ def check_type_forward(self, in_types):
     def forward(self, inputs):
         c_prev, x = inputs
         a, i, f, o = _extract_gates(x)
+        batch = len(x)
 
         if isinstance(x, numpy.ndarray):
             self.a = numpy.tanh(a)
             self.i = _sigmoid(i)
             self.f = _sigmoid(f)
             self.o = _sigmoid(o)
 
-            self.c = self.a * self.i + self.f * c_prev
-            h = self.o * numpy.tanh(self.c)
+            c_next = numpy.empty_like(c_prev)
+            c_next[:batch] = self.a * self.i + self.f * c_prev[:batch]
+            h = self.o * numpy.tanh(c_next[:batch])
         else:
-            self.c, h = cuda.elementwise(
+            c_next = cuda.cupy.empty_like(c_prev)
+            h = cuda.cupy.empty_like(c_next[:batch])
+            cuda.elementwise(
                 'T c_prev, T a, T i_, T f, T o', 'T c, T h',
                 '''
                     COMMON_ROUTINE;
                     c = aa * ai + af * c_prev;
                     h = ao * tanh(c);
                 ''',
-                'lstm_fwd', preamble=_preamble)(c_prev, a, i, f, o)
+                'lstm_fwd', preamble=_preamble)(
+                    c_prev[:batch], a, i, f, o, c_next[:batch], h)
 
-        return self.c, h
+        c_next[batch:] = c_prev[batch:]
+        self.c = c_next[:batch]
+        return c_next, h
 
     def backward(self, inputs, grad_outputs):
         xp = cuda.get_array_module(*inputs)
         c_prev, x = inputs
+        batch = len(x)
         gc, gh = grad_outputs
 
         gx = xp.empty_like(x)
         ga, gi, gf, go = _extract_gates(gx)
 
         # Consider the case that either gradient is not given
         if gc is None:
-            gc = 0
+            gc_update = 0
+            gc_rest = 0
+        else:
+            gc_update = gc[:batch]
+            gc_rest = gc[batch:]
         if gh is None:
             gh = 0
 
         if xp is numpy:
             co = numpy.tanh(self.c)
-            gc_prev = gh * self.o * _grad_tanh(co) + gc  # multiply f later
-            ga[:] = gc_prev * self.i * _grad_tanh(self.a)
-            gi[:] = gc_prev * self.a * _grad_sigmoid(self.i)
-            gf[:] = gc_prev * c_prev * _grad_sigmoid(self.f)
+            gc_prev = numpy.empty_like(c_prev)
+            # multiply f later
+            gc_prev[:batch] = gh * self.o * _grad_tanh(co) + gc_update
```

Comment:
Is it faster if we assign temporary variable to `gc_prev[:batch]`? I do not have clear answer, but I thought it because it can reduce slicing.